File: Active/C/p004.c
// Project Euler
// Problem 4: The largest palindrom made from the product of
//            3 digit numbers

#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <math.h> // README! add " -lm" when compiling this file

bool is_palindrome(unsigned long int num) {
    // convert the integer to string
    char str[256];
    sprintf(str, "%lu", num);

    // test if the string is palindrome
    int len = strlen(str);
    for (int i = 0; i < len/2; i++) {
        // the index that mirror i from the right side of the string
        int index = len - (i + 1);
        if (str[i] != str[index]) return false;
    }

    return true;
}

// generate the maximum number of n digit. if digit is 3, than the maxNum is 999
unsigned long int max_num_generator(int digit) {
    unsigned long int maxNum = 0;
    for (int i = 0; i < digit; i++)
        maxNum += 9 * (int)pow(10, i);
    return maxNum;
}

// generate the minimum number of n digit. if digit is 3, than the maxNum is 100
unsigned long int min_num_generator(int digit) {
    unsigned long int minNum = (int)pow(10, digit - 1);
    return minNum;
}

int compute(int digit) {
    unsigned long int maxNum = max_num_generator(digit);
    unsigned long int maxProduct = maxNum * maxNum;

    while (maxProduct > 0) {
        if (is_palindrome(maxProduct)) {
            int divider = maxNum;
            int minNum = min_num_generator(digit);

            while (divider >= minNum) {
                if (maxProduct % divider == 0 && maxProduct/divider <= maxNum) {
                    return maxProduct;
                }
                divider -= 1;
            }
        }
        maxProduct -= 1;
    }
    return -1;
}

int main(int argc, char *argv[]) {
    int digit = 2;
    printf("%d\n", compute(digit));
    digit = 3;
    printf("%d\n", compute(digit));

    return 0;
}

File: Active/TypeScript/src/P004/P004.ts
// Project Euler
// Problem 4: The largest palindrome made from the product of two 3-digit numbers

export function isPalindrome(num: number): boolean {
    const numStr: string = String(num);
    const numLen: number = numStr.length;

    let i = 0;
    while (i < Math.floor(numLen / 2)) {
        if (numStr[i] !== numStr[numLen - 1 -i]) {
            return false;
        }
        i += 1;
    }

    return true;
}

// generate the maximum number of n digit. if digit is 3, than the maxNum is 999
function maxNumGenerator(digit: number): number {
    let maxNum: string = "";
    for (let i = 0; i < digit; i++) maxNum += "9";
    return Number(maxNum);
}

// generate the minimum number of n digit. if digit is 3, than the maxNum is 100
function minNumGenerator(digit: number): number {
    let minNum: string = "";
    for (let i = 0; i < (digit - 1); i++) minNum += "9";
    return Number(minNum) + 1;
}

export function compute004(digit: number): number {
    const maxNum: number = maxNumGenerator(digit);
    let maxProduct: number = maxNum * maxNum;

    while (maxProduct > 0) {
        if (isPalindrome(maxProduct)) {
            let divider: number = maxNum;
            let minNum: number = minNumGenerator(digit);

            while (divider >= minNum) {
                if (maxProduct % divider === 0 && maxProduct/divider <= maxNum) {
                    return maxProduct;
                }
                divider -= 1;
            }
        }
        maxProduct -= 1;
    }
    return -1;
}

function run004(): void {
    const digit: number = 3;
    console.log(compute004(digit).toString());
}
